Sources/FoundationNetworking/URLSession/HTTP/HTTPURLProtocol.swift

```diff
@@ -499,9 +499,106 @@ internal class _HTTPURLProtocol: _NativeProtocol {
         }
         return nil
     }
-}
+    
+    /// Whenever we receive a response (i.e. a complete header) from libcurl,
+    /// this method gets called.
+    func didReceiveResponse() {
+        guard let _ = task as? URLSessionDataTask else { return }
+        guard case .transferInProgress(let ts) = self.internalState else { fatalError("Transfer not in progress.") }
+        guard let response = ts.response as? HTTPURLResponse else { fatalError("Header complete, but not URL response.") }
+        guard let session = task?.session as? URLSession else { fatalError() }
+        switch session.behaviour(for: self.task!) {
+        case .noDelegate:
+            break
+        case .taskDelegate:
+            //TODO: There's a problem with libcurl / with how we're using it.
+            // We're currently unable to pause the transfer / the easy handle:
+            // https://curl.haxx.se/mail/lib-2016-03/0222.html
+            //
+            // For now, we'll notify the delegate, but won't pause the transfer,
+            // and we'll disregard the completion handler:
+            switch response.statusCode {
+            case 301, 302, 303, 305...308:
+                break
+            default:
+                self.client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
+            }
+        case .dataCompletionHandler:
+            break
+        case .downloadCompletionHandler:
+            break
+        }
+    }
 
-fileprivate extension _HTTPURLProtocol {
+    /// If the response is a redirect, return the new request
+    ///
+    /// RFC 7231 section 6.4 defines redirection behavior for HTTP/1.1
+    ///
+    /// - SeeAlso: <https://tools.ietf.org/html/rfc7231#section-6.4>
+    func redirectRequest(for response: HTTPURLResponse, fromRequest: URLRequest) -> URLRequest? {
+        //TODO: Do we ever want to redirect for HEAD requests?
+        
+        guard
+            let location = response.value(forHeaderField: .location),
+            let targetURL = URL(string: location)
```

Comment:
Location can be relative. Use `string:relativeTo:` initializer which can handle both.

Reply:
This is existing behavior, the change here just moved `redirectRequest` out of an extension (see old line 656).  I'd prefer to split this out into a separate issue.